### PR TITLE
Teach cup to connect to Tailscale

### PR DESCRIPTION
**Pull request by brettbuddin**
- Teach cup how to connect to Tailscale.
- Upgrades minimum Go version to 1.21.
- Move from `golang.org/x/exp/slog` -> `log/slog`.
- Implements an HTTP middleware that uses `log/slog`.

**Usage:**

```
; cup --tailscale-hostname=coffee
```

I was interested in standing cup up within my Tailscale network so I added [tsnet](https://pkg.go.dev/tailscale.com/tsnet).

When the `--tailscale-hostname` is set, it will attempt to connect the service to Tailscale at that hostname. You'll need to either supply a token or visit the URL in the logs to authenticate the service within the tailnet.

The HTTP server is wrapped in a middleware that extracts identity information from all requests and supplies it to later stages via the request's `context.Context`. The identity can be extracted using the helper function: `tailscale.WhoIs(ctx)`.

### CODE REVIEW

**Comment by brettbuddin** [comment]
@GeorgeMac I'll go ahead and try and make a middleware that conforms to that. At least then `log.Printf` and `slog` will be used together and it'll look somewhat the same.

**Comment by brettbuddin** [comment]
I'll swap out `exp/slices` and friends with their first-class cousins while I'm in here.

File: pkg/api/logger/middleware.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+// New returns a middleware function that logs requests using structure logging.
+func New(handler slog.Handler) func(next http.Handler) http.Handler {
+	return middleware.RequestLogger(
+		&slogger{handler: handler},
+	)
+}
+
+// EntrySet extends request's [*slog.Logger] with additional fields. Argumennts are converted to
```
Comment (GeorgeMac) [inline pkg/api/logger/middleware.go:20]:
```suggestion
// EntrySet extends request's [*slog.Logger] with additional fields. Arguments are converted to
```

**Review by GeorgeMac — APPROVED** [review]
This is super cool. Thanks dude! 🙇

**Pull request merged**